src/MemoryRouter.tsx

```diff
@@ -165,7 +165,19 @@ export class MemoryRouter extends BaseRouter {
       query: parsedUrl.query || {},
       hash: parsedUrl.hash || "",
     };
-    const asPath = getRouteAsPath(newRoute.pathname, newRoute.query, newRoute.hash);
+
+    let asPath: string;
+    if (as === undefined) {
+      asPath = getRouteAsPath(newRoute.pathname, newRoute.query, newRoute.hash);
+    } else {
+      const parsedAsUrl = typeof as === "object" ? as : parseUrl(as, true);
+      let asRoute: UrlObjectComplete = {
+        pathname: removeTrailingSlash(parsedAsUrl.pathname ?? this.pathname),
+        query: parsedAsUrl.query || {},
+        hash: parsedAsUrl.hash || "",
+      };
```

Comment:
I would love to extract this into a function, to make this cleaner and avoid duplication.  Do you mind making that change?